engine/buildkit/client.go

```diff
@@ -285,57 +285,6 @@ func (c *Client) Solve(ctx context.Context, req bkgw.SolveRequest) (_ *Result, r
 	var llbRes *bkfrontend.Result
 	switch {
 	case req.Definition != nil && req.Definition.Def != nil:
-		clientMetadata, err := engine.ClientMetadataFromContext(ctx)
-		if err != nil {
-			return nil, err
-		}
-
-		dag, err := DefToDAG(req.Definition)
-		if err != nil {
-			return nil, err
-		}
-		if err := dag.Walk(func(dag *OpDAG) error {
-			execOp, ok := dag.AsExec()
-			if !ok {
-				return nil
-			}
-			if execOp.Meta == nil {
-				execOp.Meta = &bksolverpb.Meta{}
-			}
-			if execOp.Meta.ProxyEnv == nil {
-				execOp.Meta.ProxyEnv = &bksolverpb.ProxyEnv{}
-			}
-
-			// If we already solved for this execop in this dagger session, use
-			// it's uncached metadata to preserve the same vertex digest.
-			// This results in buildkit's solver de-duping the op, instead of
-			// the scheduler's edge merge.
```

Comment:
So this was originally added because of https://github.com/dagger/dagger/pull/6505.

Because we're not using buildkit's progress stream this shouldn't be an issue anymore right?

We still would be doing more edge merging again because of this - since these are likely what cause inconsistent graph state issues, we might see more of those (though that would definitely help debugging it lol).

If there's an easy way to keep this hack in (I like the explicitness of it, instead of hiding the details in the solver), I'd prefer that, but not a blocker.

> (edit - side note: it seems like we could potentially remove the ftp_proxy hack entirely if we replace ServerID with the OpenTelemetry trace + span ID.)

I like this - this is the best of all worlds - we get this feature, and we get rid of our nasty hack.

Reply:
> I like this - this is the best of all worlds - we get this feature, and we get rid of our nasty hack.

Yeah, I really want to do this, but it probably makes the most sense to do it after #6914 and #6916 (cc @sipsma) since at that point we for sure don't need the `ParentClientIDs` value, and solely have the `ServerID` to figure out. At that point theoretically we just stop using a random ID for it and use something analogous to `$traceparent` instead. (Carefully pointing out we need trace ID + span ID, not just trace ID, to handle nesting.)

We'd be adding a hard dependency on OpenTelemetry at this layer, but that seems worth the trade-off to me.

Reply:
Have invested 0 time looking myself because on mobile atm, but how is the otel trace id propagated? Not necessarily opposed to using it as the server id but if we can just propagate the server id the same way we propagate the trace id that might be even better (just in the interest of not creating too much entanglement).

Also, the approach I'm adding in the CA cert PR of using a custom Executor rather than the shim should in the long term let us migrate everything in the shim and also be another route of avoiding the ftp proxy hack. Just noting for completeness.

Reply:
@sipsma We can almost certainly use the new custom Executor if we want to yeah. Right now Buildkit propagates it for us with its OpenTelemetry integration - somewhere around [here](https://github.com/moby/buildkit/blob/647a997b389757068760410053873745acabfc80/executor/oci/spec.go#L125) I suppose (I never had to look).

FWIW, it seems worth considering making ServerID == TraceID+SpanID regardless. I briefly considered just doing it along the way, but had enough change going on and didn't bother. It doesn't seem like we gain a whole lot by making up a separate random string, vs. if we make it match, we can automatically correlate a ServerID to a trace for "free" which could come in handy when going through logs and such. Low conviction on that though, wanted to ask and now's my chance. :grin: